packages/eslint-plugin/tests/rules/no-unnecessary-condition.test.ts

```diff
@@ -1670,26 +1686,6 @@ pick({ foo: 1, bar: 2 }, 'bar');
     },
     {
       code: `
-function getElem(dict: Record<string, { foo: string }>, key: string) {
-  if (dict[key]) {
-    return dict[key].foo;
-  } else {
-    return '';
-  }
-}
```

Comment:
This was arguably a bug fix in the existing test case, because it shouldn't be reported either.

Reply:
This actually should be reported because the fixture config doesn't use `noUncheckedIndexAccess`!

I think we should probably guard the new code around whether or not the flag is on.

Reply:
@bradzacher This code is the exact same principle as `dictionary[key] ??= defaultVal`. If we want to ignore that, we have to ignore others as well.

Reply:
They are and they aren't. You're correct in that they're both record index access in a logical position - so they're semantically the same case. However TS reports the types differently for the two cases.

In this instance TS will report the read type, meaning that if `noUncheckedIndexAccess` is turned on then TS will report the type as `T | undefimed` to us.

However in `x ?? y`, TS will report the write type, meaning it reports just `T` - no matter whether or not `noUncheckedIndexAccess` is on or not.

This is the fundamental difference between the two cases and why we are false positiving on the assignment operators, but _not_ on this case.

Reply:
However, we already ignore array index access, even without `noUncheckedIndexAccess`. See this test case:

```ts
declare const arr: { foo: number }[];
const bar = arr[42]?.foo ?? 0;
```

I find it quite natural to extend the scope to all index signatures, since there isn't something special about array indices.

Reply:
The reason we did it for arrays and not objects is because defining arrays as `(T | undefined)[]` is similar to `Record<string, T | undefined>`, but far, far less ergonomic syntactically and is arguably worse given how common it is to iterate arrays directly using an iterator method or for/of.

For objects it's pretty natural to union in `undefined` because most usecases for reading is an index signature - the direct iteration methods (`Object.entries` / `Object.values`) are both rarer usecases so the `undefined` doesn't get in the way as much.

Now that we have `noUncheckedIndexAccess` - we don't really need the logic at all, but I understand that some people think the compiler option goes too far and can get in the way (which is true to a degree).

The motivation behind this change isn't bringing the two sets up to parity (that's a much larger change that we should discuss as a group to decide if we think we should treat objects the same as arrays in this regard) - instead it's just fixing the current false positives we have due to the logical operator handling we added in #6594